File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/messaging/MessagingCaptor.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.messaging;

import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.interceptor.captor.common.PropertyServiceNameDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.convert.TypeConverter;
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever;
import io.lsdconsulting.lsd.distributed.interceptor.persistance.RepositoryService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lsd.format.PrettyPrinter;
import org.springframework.messaging.Message;

import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Collection;
import java.util.Map;

import static io.lsdconsulting.lsd.distributed.access.model.InteractionType.CONSUME;
import static io.lsdconsulting.lsd.distributed.access.model.InteractionType.PUBLISH;
import static io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver.SOURCE_NAME_KEY;
import static io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver.TARGET_NAME_KEY;
import static java.util.Collections.emptyMap;
import static org.apache.commons.lang3.StringUtils.isBlank;

@RequiredArgsConstructor
@Slf4j
public class MessagingCaptor {

    private final RepositoryService repositoryService;
    private final PropertyServiceNameDeriver propertyServiceNameDeriver;
    private final TraceIdRetriever traceIdRetriever;
    private final MessagingHeaderRetriever messagingHeaderRetriever;
    private final String profile;

    public InterceptedInteraction captureConsumeInteraction(final Message<?> message) {
        Map<String, Collection<String>> headers = messagingHeaderRetriever.retrieve(message);
        final InterceptedInteraction interceptedInteraction = new InterceptedInteraction(
                traceIdRetriever.getTraceId(headers),
                PrettyPrinter.prettyPrint(TypeConverter.convert((byte[]) message.getPayload())),
                headers,
                emptyMap(),
                propertyServiceNameDeriver.getServiceName(),
                getSource(message),
                propertyServiceNameDeriver.getServiceName(),
                null,
                null,
                CONSUME,
                profile,
                0L,
                ZonedDateTime.now(ZoneId.of("UTC")));

        repositoryService.enqueue(interceptedInteraction);
        return interceptedInteraction;
    }

    private String getSource(Message<?> message) {
        String source = (String) message.getHeaders().get(TARGET_NAME_KEY);
        if (isBlank(source)) {
            String typeIdHeader = (String) message.getHeaders().get("__TypeId__");
            source = getSourceFrom(typeIdHeader);
        }
        log.debug("found source:{}", source);
        return source;
    }

    private String getSourceFrom(String typeIdHeader) {
        String source = "UNKNOWN";
        if (!isBlank(typeIdHeader)) {
            String[] sourceTokens = typeIdHeader.split("\\.");
            source = sourceTokens[sourceTokens.length - 1];
        }
        return source;
    }

    public InterceptedInteraction capturePublishInteraction(final Message<?> message) {
        String source = (String) message.getHeaders().get(SOURCE_NAME_KEY);
        String target = (String) message.getHeaders().get(TARGET_NAME_KEY);

        Map<String, Collection<String>> headers = messagingHeaderRetriever.retrieve(message);
        final InterceptedInteraction interceptedInteraction = new InterceptedInteraction(
                traceIdRetriever.getTraceId(headers),
                TypeConverter.convert((byte[]) message.getPayload()),
                headers,
                emptyMap(),
                source != null ? source : propertyServiceNameDeriver.getServiceName(),
                target,
                target,
                null,
                null,
                PUBLISH,
                profile,
                0L,
                ZonedDateTime.now(ZoneId.of("UTC")));

        repositoryService.enqueue(interceptedInteraction);
        return interceptedInteraction;
    }
}

File: spring-messaging/src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/messaging/MessagingCaptor.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.messaging

import io.lsdconsulting.lsd.distributed.connector.model.InteractionType.CONSUME
import io.lsdconsulting.lsd.distributed.connector.model.InteractionType.PUBLISH
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.interceptor.captor.common.PropertyServiceNameDeriver
import io.lsdconsulting.lsd.distributed.interceptor.captor.common.print
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.SourceTargetDeriver.Companion.SOURCE_NAME_KEY
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.SourceTargetDeriver.Companion.TARGET_NAME_KEY
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever
import io.lsdconsulting.lsd.distributed.interceptor.config.log
import io.lsdconsulting.lsd.distributed.interceptor.persistence.RepositoryService
import org.springframework.messaging.Message
import java.time.ZoneId
import java.time.ZonedDateTime

class MessagingCaptor(
    private val repositoryService: RepositoryService,
    private val propertyServiceNameDeriver: PropertyServiceNameDeriver,
    private val traceIdRetriever: TraceIdRetriever,
    private val messagingHeaderRetriever: MessagingHeaderRetriever,
    private val profile: String,
) {
    fun captureConsumeInteraction(message: Message<*>): InterceptedInteraction {
        val headers = messagingHeaderRetriever.retrieve(message)
        val interceptedInteraction = InterceptedInteraction(
            traceId = traceIdRetriever.getTraceId(headers),
            body = print(message.payload),
            requestHeaders = headers,
            responseHeaders = emptyMap(),
            serviceName = propertyServiceNameDeriver.serviceName,
            target = getTarget(message),
            path = propertyServiceNameDeriver.serviceName,
            httpStatus = null, httpMethod = null,
            interactionType = CONSUME, profile = profile, elapsedTime = 0L,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        repositoryService.enqueue(interceptedInteraction)
        return interceptedInteraction
    }

    // TODO Add support for info.app.name
    private fun getTarget(message: Message<*>): String {
        val header = message.headers[TARGET_NAME_KEY]
        var target = if (header is String) header else if (header is ByteArray) String(header) else null
        if (target.isNullOrEmpty()) {
            val typeIdHeader = message.headers["__TypeId__"] as String?
            target = getTargetFrom(typeIdHeader)
        }
        if (target.isNullOrEmpty()) {
            target = "UNKNOWN"
        }
        log().debug("found target:{}", target)
        return target
    }

    private fun getTargetFrom(typeIdHeader: String?): String? {
        return if (!typeIdHeader.isNullOrBlank()) {
            return typeIdHeader.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray<String>().last()
        }
        else null
    }

    fun capturePublishInteraction(message: Message<*>): InterceptedInteraction {
        val source = message.headers[SOURCE_NAME_KEY] as String?
        val target = message.headers[TARGET_NAME_KEY] as String
        val headers = messagingHeaderRetriever.retrieve(message)
        val interceptedInteraction = InterceptedInteraction(
            traceId = traceIdRetriever.getTraceId(headers),
            body = print(message.payload),
            requestHeaders = headers,
            responseHeaders = emptyMap(),
            serviceName = source ?: propertyServiceNameDeriver.serviceName,
            target = target, path = target,
            httpStatus = null, httpMethod = null,
            interactionType = PUBLISH, profile = profile, elapsedTime = 0L,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        repositoryService.enqueue(interceptedInteraction)
        return interceptedInteraction
    }
}
